src/java/org/apache/cassandra/cache/ChunkCache.java

```diff
@@ -158,88 +212,260 @@ public boolean equals(Object obj)
+         *
+         * @param position the position requested by the client, this has not been aligned to neither the chunk nor the page
+         * @return the buffer covering this position, or null if no reference can be taken
+         */
+        @Nullable
+        Rebufferer.BufferHolder getReferencedBuffer(long position)
         {
             int refCount;
             do
             {
-                refCount = references.get();
+                refCount = references;
+
                 if (refCount == 0)
-                    // Buffer was released before we managed to reference it.
-                    return null;
-            } while (!references.compareAndSet(refCount, refCount + 1));
+                    return null; // Buffer was released before we managed to reference it.
 
-            return this;
+            } while (!referencesUpdater.compareAndSet(this, refCount, refCount + 1));
+
+            return getBuffer(position);
         }
 
-        @Override
-        public ByteBuffer buffer()
+        /**
+         * Release the chunk when the cache or a reader no longer needs it.
+         */
+        public void release()
         {
-            assert references.get() > 0;
-            return buffer.duplicate();
+            if (referencesUpdater.decrementAndGet(this) == 0)
+                releaseBuffers();
         }
 
-        @Override
-        public ByteOrder order()
+        public long offset()
         {
-            return buffer.order();
+            return offset;
         }
 
-        @Override
-        public FloatBuffer floatBuffer()
+        /**
+         * Load the data for this chunk.
+         *
+         * @param key The key / file & position that this chunk corresponds to.
+         * @return a future holding the loaded chunk when it completes.
+         */
+        abstract void read(Key key);
+
+        /**
+         * Return the correct buffer depending on the position requested by the client. The returned buffer cannot be
+         * null, but may be empty and must contain the given position
+         * (i.e. buf.offset <= position <= buf.offset + buf.buffer.limit where the latter can only be == if at the end
+         * of the file and buffer is empty).
+         *
+         * @param position the position requested by the client, this has not been aligned to neither the chunk nor the page
+         * @return the buffer covering this position, or null if no reference can be taken
+         */
+        abstract Rebufferer.BufferHolder getBuffer(long position);
+
+        /**
+         * @return the space taken by this chunk
+         */
+        abstract int capacity();
+
+        /**
+         * Release the addresses when this chunk is no longer used. Called when all references are released.
+         */
+        abstract void releaseBuffers();
+    }
+
+    private static final AtomicIntegerFieldUpdater<Chunk> referencesUpdater = AtomicIntegerFieldUpdater.newUpdater(Chunk.class, "references");
```

Comment:
nit: what about moving this static field at the top of the class ?

Reply:
It's a field of `Chunk`, moved there (Java 8 did not permit statics in nested classes).